pd/src/state/writer.rs

```diff
@@ -161,30 +179,181 @@ impl Writer {
+
+            // Keep track of the note so we can insert that as well for the sake of CompactBlock
+            // it will need the position the note commitment tree
+            let position = note_commitment_tree
+                .bridges()
+                .last()
+                .map(|b| b.frontier().position().into())
+                // If there are no bridges, the tree is empty
+                .unwrap_or(0u64);
+            let positioned_note = PositionedNoteData {
+                position,
+                data: note_data,
+            };
+
+            notes.push((commitment, positioned_note));
+        }
+
+        // We might not have any allocations of some delegation tokens, but we should record the denoms.
+        for genesis::ValidatorPower { validator, .. } in genesis_config.validators.iter() {
+            let denom = validator.identity_key.delegation_token().denom();
+            supply_updates.entry(denom.id()).or_insert((denom, 0)).1 += 0;
+        }
+
+        // update the NCT
+        let nct_anchor = note_commitment_tree.root2();
+        let (jmt_root, tree_update_batch) = jmt::JellyfishMerkleTree::new(&self.private_reader)
+            .put_value_set(
+                // TODO: create a JmtKey enum, where each variant has
+                // a different domain-separated hash
+                vec![(
+                    jellyfish::Key::NoteCommitmentAnchor.hash(),
+                    nct_anchor.clone(),
+                )],
+                // height 0 for genesis
+                0,
+            )
+            .await?;
+        // ... and then write the resulting batch update to the backing store:
+        jellyfish::DbTx(&mut dbtx)
+            .write_node_batch(&tree_update_batch.node_batch)
+            .await?;
+
+        // The app hash needs to be returned to Tendermint
+        let app_hash: [u8; 32] = jmt_root.to_vec().try_into().unwrap();
 
-            next_rate_data.insert(
-                validator.identity_key.clone(),
-                penumbra_stake::RateData {
-                    identity_key: validator.identity_key.clone(),
-                    epoch_index: 1,
-                    validator_reward_rate: 0,
-                    validator_exchange_rate: 1_0000_0000,
-                },
-            );
+        // Insert the block into the DB
+        query!(
+            "INSERT INTO blocks (height, nct_anchor, app_hash) VALUES ($1, $2, $3)",
+            0 as i64,
+            &nct_anchor.to_bytes()[..],
+            &app_hash[..]
+        )
+        .execute(&mut dbtx)
+        .await?;
+
+        // write the token supplies to the database
+        for (id, asset) in &supply_updates {
+            query!(
+                "INSERT INTO assets (asset_id, denom, total_supply)
+            VALUES ($1, $2, $3)
+            ON CONFLICT (asset_id) DO UPDATE SET denom=$2, total_supply=$3",
+                &id.to_bytes()[..],
+                asset.0.to_string(),
+                asset.1 as i64
```

Comment:
ditto on unchecked cast here: if we have a total supply close to max u64 this could overflow